gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedMultiDatasets.java

```diff
@@ -32,54 +34,101 @@
-      // which imply that we need to at least have a distcp-ng job per copyTo cluster
-      if(rc.getCopyMode()== ReplicationCopyMode.PUSH){
-        log.warn("Currently not support for push mode of dataset with metadata {}", rc.getMetaData());
-        return;
+      // push mode
+      if(this.props.containsKey(REPLICATION_PUSH_MODE) && Boolean.parseBoolean(this.props.getProperty(REPLICATION_PUSH_MODE))){
+        generateDatasetInPushMode(rc, executionClusterURI);
       }
-      
-      // PULL mode
-      CopyRouteGenerator cpGen = rc.getCopyRouteGenerator();
-      List<EndPoint> replicas = rc.getReplicas();
-      for(EndPoint replica: replicas){
-        if(needGenerateCopyEntity(replica, executionClusterURI)){
-          Optional<CopyRoute> copyRoute = cpGen.getPullRoute(rc, replica);
-          if(copyRoute.isPresent()){
-            this.datasets.add(new ConfigBasedDataset(rc, this.props, copyRoute.get()));
-          }
-        }
+      // default pull mode
+      else{
+        generateDatasetInPullMode(rc, executionClusterURI);
       }
-
-      
     } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
       log.error("Can not create Replication Configuration from raw config "
           + c.root().render(ConfigRenderOptions.defaults().setComments(false).setOriginComments(false)), e);
     } catch (IOException ioe) {
       log.error("Can not decide current execution cluster ", ioe);
-      
+
     }
   }
-  
+
+  private void generateDatasetInPushMode(ReplicationConfiguration rc, URI executionClusterURI){
+    if(rc.getCopyMode()== ReplicationCopyMode.PULL){
+      return;
+    }
+
+    if (!this.props.containsKey(ConfigurationKeys.WRITER_FILE_SYSTEM_URI)){
+      return;
+    }
+    String pushModeTargetCluster = this.props.getProperty(ConfigurationKeys.WRITER_FILE_SYSTEM_URI);
+    
+    // PUSH mode
+    CopyRouteGenerator cpGen = rc.getCopyRouteGenerator();
+    List<EndPoint> replicas = rc.getReplicas();
+    List<EndPoint> pushCandidates = new ArrayList<EndPoint>(replicas);
+    pushCandidates.add(rc.getSource());
+    
+    for(EndPoint pushFrom: pushCandidates){
+      if(needGenerateCopyEntity(pushFrom, executionClusterURI)){
+        Optional<List<CopyRoute>> copyRoutes = cpGen.getPushRoutes(rc, pushFrom);
+        if(!copyRoutes.isPresent()) return;
+
+        for(CopyRoute cr: copyRoutes.get()){
+          if(cr.getCopyTo() instanceof HadoopFsEndPoint){
+
+            HadoopFsEndPoint ep = (HadoopFsEndPoint)cr.getCopyTo();
+            if(ep.getFsURI().toString().equals(pushModeTargetCluster)){
+              this.datasets.add(new ConfigBasedDataset(rc, this.props, cr));
+            }
+          }
+        }// inner for loops ends
+      }
+    }// outer for loop ends
+  }
+
+  private void generateDatasetInPullMode(ReplicationConfiguration rc, URI executionClusterURI){
+    if(rc.getCopyMode()== ReplicationCopyMode.PUSH){
```

Comment:
Ok. In that case, can you just some logging to note why the dataset was skipped.